pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,529 @@
+                    continue
+                expr = expr.diff(sym_vec[var_idx], nderivs)
+            replace_dict = dict(
+                (k, v) for k, v in zip(sym_vec, rand[:, rand_vec_idx])
+            )
+            eval_expr = evalf(expr.xreplace(replace_dict))
+            row.append(eval_expr)
+        row.append(1)
+        mat.append(row)
+
+    mat = sym.Matrix(mat)
+    failed = False
+    try:
+        L, U, perm = mat.LUdecomposition()
+    except RuntimeError:
+        # symengine throws an error when rank deficient
+        # and sympy returns U with last row zero
+        failed = True
+
+    if not sym.USE_SYMENGINE and all(expr == 0 for expr in U[-1, :]):
+        failed = True
+
+    if failed:
+        if retries == 0:
+            raise RuntimeError("Failed to find a base kernel")
+        return _get_base_kernel_matrix(
+            base_kernel=base_kernel,
+            hashable_kernel_arguments=hashable_kernel_arguments,
+            order=order,
+            retries=retries-1,
+        )
+
+    return (L, U, perm), rand, mis
+
+
+def evalf(expr, prec=100):
+    """evaluate an expression numerically using ``prec``
+    number of bits.
+    """
+    from sumpy.symbolic import USE_SYMENGINE
+    if USE_SYMENGINE:
+        return expr.n(prec=prec)
+    else:
+        import sympy
+        dps = int(sympy.log(2**prec, 10))
+        return expr.n(n=dps)
+
+
+def filter_kernel_arguments(knls, kernel_arguments):
+    """From a dictionary of kernel arguments, filter out arguments
+    that are not needed for the kernels given as a list and return a new
+    dictionary.
+    """
+    kernel_arg_names = set()
+
+    for kernel in knls:
+        for karg in (kernel.get_args() + kernel.get_source_args()):
+            kernel_arg_names.add(karg.loopy_arg.name)
+
+    return {k: v for (k, v) in kernel_arguments.items() if k in kernel_arg_names}
+
+# }}}
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+    from sumpy.kernel import (StokesletKernel, BiharmonicKernel,  # noqa:F401
+        StressletKernel, ElasticityKernel, LaplaceKernel)
+    base_kernel = BiharmonicKernel(2)
+    #base_kernel = LaplaceKernel(3)
+    kernels = [StokesletKernel(2, 0, 1), StokesletKernel(2, 0, 0)]
+    kernels += [StressletKernel(2, 0, 0, 0), StressletKernel(2, 0, 0, 1),
+            StressletKernel(2, 0, 0, 1), StressletKernel(2, 0, 1, 1)]
+    get_deriv_relation(kernels, base_kernel, tol=1e-10, order=4, kernel_arguments={})
+
+    base_kernel = BiharmonicKernel(3)
+    sym_d = make_sym_vector("d", base_kernel.dim)
+    sym_r = sym.sqrt(sum(a**2 for a in sym_d))
+    conv = SympyToPymbolicMapper()
+    expression_knl = ExpressionKernel(3, conv(sym_d[0]*sym_d[1]/sym_r**3), 1, False)
+    expression_knl2 = ExpressionKernel(3, conv(1/sym_r + sym_d[0]*sym_d[0]/sym_r**3),
+        1, False)
+    kernels = [expression_knl, expression_knl2]
+    get_deriv_relation(kernels, base_kernel, tol=1e-10, order=4, kernel_arguments={})
```

Comment:
Is this stuff being run by `pytest`? Maybe move it to a test if it doesn't :D

Reply:
This is just for me to visualize things while debugging.

Reply:
Ah, fair enough, if this functionality is already tested elsewhere.